code/datums/components/decal.dm

```diff
@@ -1,100 +1,78 @@
-
-/datum/component/decal/UnregisterFromParent()
-	UnregisterSignal(parent, list(COMSIG_ATOM_DIR_CHANGE, COMSIG_COMPONENT_CLEAN_ACT, COMSIG_PARENT_EXAMINE))
-
-/datum/component/decal/Destroy()
-	remove()
-	return ..()
-
-/datum/component/decal/PreTransfer()
-	remove()
-
-/datum/component/decal/PostTransfer()
-	remove()
-	apply()
-
-/datum/component/decal/proc/generate_appearance(_icon, _icon_state, _dir, _layer, _color, _alpha)
+	target.add_overlay(pic)
+	RegisterSignal(target,COMSIG_ATOM_UPDATE_OVERLAYS,.proc/apply_overlay, TRUE)
+	if(target.flags_1 & INITIALIZED_1)
+		target.update_icon() //could use some queuing here now maybe.
+	else
+		RegisterSignal(target,COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZE,.proc/late_update_icon, TRUE)
+	if(isitem(target))
+		INVOKE_ASYNC(target, /obj/item/.proc/update_slot_icon, TRUE)
+	if(_dir)
+		RegisterSignal(target, COMSIG_ATOM_DIR_CHANGE, .proc/rotate_react,TRUE)
+	if(_cleanable)
+		RegisterSignal(target, COMSIG_COMPONENT_CLEAN_ACT, .proc/clean_react,TRUE)
+	if(_description)
+		RegisterSignal(target, COMSIG_PARENT_EXAMINE, .proc/examine,TRUE)
+
+/datum/element/decal/proc/generate_appearance(_icon, _icon_state, _dir, _layer, _color, _alpha, source)
 	if(!_icon || !_icon_state)
 		return FALSE
-	// It has to be made from an image or dir breaks because of a byond bug
 	var/temp_image = image(_icon, null, _icon_state, _layer, _dir)
 	pic = new(temp_image)
 	pic.color = _color
 	pic.alpha = _alpha
 	return TRUE
 
-/datum/component/decal/proc/apply(atom/thing)
-	var/atom/master = thing || parent
-	RegisterSignal(master,COMSIG_ATOM_UPDATE_OVERLAYS,.proc/apply_overlay)
-	if(master.flags_1 & INITIALIZED_1)
-		master.update_icon() //could use some queuing here now maybe.
-	else
-		RegisterSignal(master,COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZE,.proc/late_update_icon)
-	if(isitem(master))
-		addtimer(CALLBACK(master, /obj/item/.proc/update_slot_icon), 0, TIMER_UNIQUE)
+/datum/element/decal/Detach(atom/source, force)
+	UnregisterSignal(source, list(COMSIG_ATOM_DIR_CHANGE, COMSIG_COMPONENT_CLEAN_ACT, COMSIG_PARENT_EXAMINE))
+	UnregisterSignal(source,COMSIG_ATOM_UPDATE_OVERLAYS)
+	source.update_icon()
+	if(isitem(source))
+		INVOKE_ASYNC(source, /obj/item/.proc/update_slot_icon)
+	return ..()
 
-/datum/component/decal/proc/late_update_icon(atom/source)
+/datum/element/decal/proc/late_update_icon(atom/source)
 	SIGNAL_HANDLER
 
 	if(source && istype(source))
 		source.update_icon()
 		UnregisterSignal(source,COMSIG_ATOM_AFTER_SUCCESSFUL_INITIALIZE)
 
-/datum/component/decal/proc/apply_overlay(atom/source, list/overlay_list)
+/datum/element/decal/proc/apply_overlay(atom/source, list/overlay_list)
 	SIGNAL_HANDLER
 
-	overlay_list += pic
-
-/datum/component/decal/proc/remove(atom/thing)
-	var/atom/master = thing || parent
-	UnregisterSignal(master,COMSIG_ATOM_UPDATE_OVERLAYS)
-	master.update_icon()
-	if(isitem(master))
-		addtimer(CALLBACK(master, /obj/item/.proc/update_slot_icon), 0, TIMER_UNIQUE)
+	source.add_overlay(pic)
```

Comment:
Should be `overlay_list += pic`
Overlay signal should not be adding the overlays itself, instead it should return them in overlay list.